src/common/async.cpp

```diff
@@ -7,138 +7,58 @@
-        if (!there) throw std::runtime_error("Element not there");             \
-        v = m[k];                                                              \
-    } while (0)
-
-void Async::loop_thread(Var<AsyncState> state) {
-    state->interrupted = false; // Undo previous break_loop() if any
-
-    debug("async: loop thread entered");
-    for (;;) {
-
-        LOCKED({
-            debug("async: loop thread locked");
-            auto empty = (state->ready.empty() && state->active.empty());
-            if (state->interrupted || empty) {
-                debug("async: %s", (empty) ? "empty" : "interrupted");
-                // Do this here so it's done while we're locked
-                debug("async: detaching thread...");
-                state->thread.detach();
-                state->thread_running = false;
-                break;
-            }
-
-            debug("async: not interrupted and not empty");
-            for (auto pair : state->ready) {
-                NetTest *ptr = pair.first;
-                debug("async: active: %lld", ptr->identifier());
-                INSERT(state->active, ptr, pair.second);
-                // Note: it should be safe to capture `state` by reference
-                // here because the destructor should not be able to destroy
-                // the background thread until we detach it and break.
-                ptr->begin([&state, ptr]() {
-                    ptr->end([&state, ptr]() {
-                        //
-                        // This callback is invoked by loop_once, when we do
-                        // not own the lock. For this reason it's important
-                        // to only modify state->active in the current thread,
-                        // i.e. in the background thread (i.e this function)
-                        //
-                        debug("async: test stopped");
-                        Var<NetTest> test;
-                        std::function<void(Var<NetTest>)> callback;
-                        LOCKED({
-                            GET(test, state->active, ptr);
-                            GET(callback, state->callbacks, ptr);
-                            state->active.erase(ptr);
-                            state->callbacks.erase(ptr);
-                            // Keep alive test until end of the loop
-                            INSERT(state->complete, ptr, test);
-                        });
-                        if (callback) {
-                            callback(test);
-                        }
-                    });
+    active += 1;
+    debug("async: scheduling %llu", test->identifier());
+    reactor->call_later(1.0, [=]() {
+        debug("async: starting %llu", test->identifier());
+        test->begin([=]() {
+            debug("async: ending %llu", test->identifier());
+            test->end([=]() {
+                debug("async: cleaning-up %llu", test->identifier());
+                reactor->call_soon([=]() {
```

Comment:
I should probably add a comment to explain that, because otherwise it's not super clear. @AntonioLangiu what do you think?